New presets locations

sophiejandrews opened the pull request.

sophiejandrews committed:
add break even buttons to proposed sliders

sophiejandrews committed:
cleaned up code a bit for break even buttons

sophiejandrews committed:
moved preset menus to box at top

sophiejandrews committed:
fixed layout of preset menus at top

sophiejandrews committed:
UI changes to preset selections at top

sophiejandrews committed:
added ILR presetand two preset buttons

sophiejandrews committed:
cleaned up code

sophiejandrews committed:
non linear ILR slider

mdeceglie commented:
The new presets interface looks great! Here are a few items to polish before we merge:

1. Switch the order of ILR and load buttons so that when the window is narrow, the load buttons appear on the bottom. (thanks for catching this @sophiejandrews)
2. Some break-even button artifacts remain. Let's either get rid of those or merge #7 first and update this.
3. Add ticks (without numbers) to the ILR slider since it can only take on discrete values

sophiejandrews committed:
switched location of load buttons

sophiejandrews committed:
switch buttons location, added ticks to ILR slider

silverman commented:
ILR is inverter load*ing* ratio.

Let's make the text box for ILR read-only so I can't get confused when I enter an unavailable ILR.

For the record: the ILR control doesn't currently do anything because we haven't calculated a new library of presets with different ILRs.

Any thoughts on changing "Load for baseline" to "Apply to baseline"? When I click a button that says "load" I half-expect to be prompted to select a local file to upload.

sophiejandrews committed:
read only ILR input, fixed ILR and button wording

silverman reviewed (approved):
Looking good.

mdeceglie commented:
This looks great. We should hold off on merging the ILR slider itself until we have those presets computed. Would it be better to sit on this whole PR, or to merge the rest of the preset interface changes and move the ILR slider and associated presets to a separate PR?

I like how the discrete ILR slider turned out though!

sophiejandrews committed:
updated backend for building preset tree

sophiejandrews committed:
discount rate slider on both sides

sophiejandrews committed:
added option to sync/unsync discount rate sliders

sophiejandrews committed:
checked for synchronized moment

sophiejandrews committed:
sync when box goes from unchecked to checked

sophiejandrews committed:
calculation also updates with check/uncheck

sophiejandrews committed:
new icons for sync/unsync discount rate

sophiejandrews committed:
change cursor on hover

sophiejandrews committed:
merged in release_candidate

sophiejandrews committed:
merged release_candidate, linear degradation rate

sophiejandrews committed:
testing rounding errors

sophiejandrews committed:
conditional to avoid rounding errors

sophiejandrews committed:
preset tree code w/ comments

sophiejandrews committed:
relative file paths for preset tree

sophiejandrews committed:
non-final service life changes

sophiejandrews committed:
service life/degradation restricted to have positive energy

sophiejandrews committed:
new BOS cost script and tree

sophiejandrews committed:
updated preset tree and script

sophiejandrews committed:
add script to make API calls

sophiejandrews committed:
fixed typo

sophiejandrews committed:
efficiency capped at 100%

sophiejandrews committed:
removed tilt from preset tree

sophiejandrews committed:
temporary slider fixes

sophiejandrews committed:
restrict degradation and service life sliders

sophiejandrews committed:
front end reads new preset trees

sophiejandrews committed:
small cleaning to html and preset script

sophiejandrews committed:
default ILR based on system type

sophiejandrews committed:
update preset instructions

sophiejandrews committed:
remove energy spreadsheet

sophiejandrews committed:
removed old preset trees

sophiejandrews committed:
update read me and requirements

sophiejandrews committed:
update read me

sophiejandrews committed:
locations in alphabetical order

sophiejandrews committed:
indentation and comments in js file

sophiejandrews committed:
fixed spacing

sophiejandrews committed:
cell technology popover

sophiejandrews committed:
BOS costs only power constrained

sophiejandrews committed:
update preset instructions

silverman-nrel committed:
add tooltips to the discount rate 'link' buttons

silverman-nrel committed:
minor wording/punctuation adjustments

sophiejandrews committed:
service life/degradation scaling for copy and apply

sophiejandrews committed:
cell technology description update

mdeceglie commented:
Wow, this PR has gotten big! @silverman or @sophiejandrews  can you please provide a brief high-level summary for what these changes encompass?

sophiejandrews commented:
Major changes on this branch:

- Everything presets related (redoing the layout, adding the ILR slider, more system types, new preset tree backend with PySAM and updated numbers, new BOS cost tree based on Vignesh's numbers)
- split discount rate slider
- restrict degradation rate and service life sliders
- new read me

sophiejandrews committed:
license on preset trees

mdeceglie commented:
The coupled movement of the degradation rate and service life sliders is surprising, is that the logical conclusion of all the constraints we discussed?

sophiejandrews commented:
I believe it should be. As one slider moves, the maximum of the other slider is restricted by the relationship `(n-0.5) * degradation < 1`. Since the maximum is changing but the slider length stays the same, it creates the appearance of coupled movement.

mdeceglie commented:
Based on our phone call, here is a proposed path on item 1: Let's use a dead area on the slider, and change the dead area rather than the position of the slider itself. The right edge of the whole slider will represent max of 5% degradation and 100 years. We should briefly look into graying out the dead area.

sophiejandrews committed:
added dead zone to degradation and service life sliders

sophiejandrews committed:
pips for dead zone of slider

mdeceglie commented on README.md:
```diff
@@ -2,7 +2,7 @@
 
 # Building a Preset Tree for NREL PVLCOE Calculator
-In the /build-presets/ folder, the MakePresetTree.py file builds a javascript file which supplies the preset values that appear in the NREL PVLCOE calculator at http://pvlcoe.nrel.gov/, for the following parameters:
+In the `/build-presets/` folder, the `MakePresetTree.py` file builds a javascript file which supplies the preset values that appear in the NREL PVLCOE calculator at http://pvlcoe.nrel.gov/, for the following parameters:
```
This list isn't quite exhaustive. I suggest just simply saying the file builds a javascipt file which supples the preset values that appear in the calculator, rather than listing them.

mdeceglie commented on build-presets/MakePresetTree.py:
```diff
@@ -1,86 +1,127 @@
+					}
+
+# print(preset_tree)			
+with open('../js/PresetTree.js', 'w') as file:
```
This may not be platform independent. Along with `from pathlib import Path` up top, my untested suggestion is:
```suggestion
with open(Path('../js/PresetTree.js'), 'w') as file:
```

mdeceglie committed:
pep8 updates to MakePresetTree.py

mdeceglie commented on build-presets/MakePresetTree.py:
```diff
@@ -1,86 +1,131 @@
-df_yield['Location'] = [label.replace(' (TMY2)', '').replace(' (TMY3)', '') for label in df_yield['Location']]
+# creating PySAM model with default info from json file (that doesn't have location info)
+json_file = open("pvwatts_inputs.json")
+dict = json.load(json_file)
```
It's good practice to avoid reassigning built in types like `dict` we should use `pvwatts_dict` or something instead for this variable

mdeceglie commented on build-presets/MakeBOSTree.py:
```diff
@@ -0,0 +1,48 @@
+	'roof-mounted, residential scale': pd.read_csv('BOS_cost_data/roof_residential.csv'),
+	'roof-mounted, commercial scale': pd.read_csv('BOS_cost_data/roof_commercial.csv'),
+	'fixed tilt, commercial scale': pd.read_csv('BOS_cost_data/fixed_tilt_commercial.csv')
+}
```
We should use platform-independent paths here

mdeceglie commented on build-presets/MakeBOSTree.py:
```diff
@@ -0,0 +1,48 @@
+	'fixed tilt, commercial scale': pd.read_csv('BOS_cost_data/fixed_tilt_commercial.csv')
+}
+
+ilr = [1.1, 1.3, 1.4]
```
ilr are separately defined here and in `MakePresetTree.py` what happens if they aren't aligned?

sophiejandrews replied:
it’s on the programmer to make sure the values are the same. but i don’t want to combine the trees since the BOS cost fit runs 15 times to make the BOS tree and the preset code runs pysam 3300 times to creates its presets. it seems silly to run the BOS cost fit 3285 unnecessary times, especially given how long the preset code takes to run already. I can leave a warning in the code that these values need to be the same between files

mdeceglie commented on build-presets/MakeBOSTree.py:
```diff
@@ -0,0 +1,48 @@
+            		)
+        	)
+			
+with open('../js/BOSCostTree.js', 'w') as file:
```
we should use a platform-independent path here

mdeceglie commented on build-presets/MakeBOSTree.py:
```diff
@@ -0,0 +1,48 @@
+start = 0.1
+while (start < 0.41):
+	efficiencies.append(round(start, 2)) 
+	start += 0.01
```
I'd suggest this more concise way to build the array. If a numpy array won't work with the curve_fit call below, can always change to a `list` first
```suggestion
efficiencies = np.arange(0.1,0.4,0.01)
```

mdeceglie commented on build-presets/get_weather_files.py:
```diff
@@ -0,0 +1,51 @@
+import pandas as pd
+import numpy as np
+import sys, os
+from IPython.display import display
```
I don't think any of these get used

mdeceglie commented on build-presets/get_weather_files.py:
```diff
@@ -0,0 +1,51 @@
+	info = pd.read_csv(url)
+	print(lat)
+	# save the info to a csv, the weather_files folder must already be created
+	info.to_csv('weather_files/'+ state + '_' +str(lat)+'_'+str(lon)+'_tmy.csv', index=False)
```
We should use a platform-independent path here

sophiejandrews committed:
improvements to js readability

mdeceglie commented on build-presets/MakePresetTree.py:
```diff
@@ -1,86 +1,131 @@
+            for ilr in inverter_loading_ratio:
+                if ilr not in preset_tree[cell_technology][package_type][system_type]:
+                    preset_tree[cell_technology][package_type][system_type][ilr] = {}
+                for solar_resource_file in weather_data:
```
`solar_resource_dict` would be a more descriptive name for this variable

mdeceglie commented on build-presets/MakePresetTree.py:
```diff
@@ -1,86 +1,131 @@
+        'roof-mounted, residential scale': 25, 'roof-mounted, commercial scale': 10, 'fixed tilt, commercial scale': 10}
 
+# 0: fixed rack, 1: fixed roof, 2: 1 axis; needed for running PySAM
+array_type = {'fixed tilt, utility scale': 0, 'single-axis tracked, utility scale': 2,
```
I think we probably want to use back-tracking for single-axis-tracked systems. This would be `array_type` 3
```suggestion
array_type = {'fixed tilt, utility scale': 0, 'single-axis tracked, utility scale': 3,
```

mdeceglie replied:
@silverman reminded me that we probably don't want back-tracking for CdTe though!

sophiejandrews replied:
so is the final conclusion that this value should be changed to 3?

mdeceglie commented on build-presets/MakePresetTree.py:
```diff
@@ -1,86 +1,131 @@
-system_yield = {k: v[0][1] for k, v in seq_keys.items()}
-degradation_rates = {k: v[0][2] for k, v in seq_keys.items()}
+# tilt angles reported in degrees, needed for running PySAM
+tilt = {'fixed tilt, utility scale': 33, 'single-axis tracked, utility scale': 33,
```
single-axis tracked should have tilt=0
```suggestion
tilt = {'fixed tilt, utility scale': 33, 'single-axis tracked, utility scale': 0,
```

mdeceglie reviewed (commented):
This is getting close, thanks @sophiejandrews ! 

I left some feedback throughout. In addition I'm having trouble reconciling our preset yields with the PVWatts calculations. For example, in Las Vegas, for a fixed tilt utility system with ILR=1.1 our preset tree has a yield of about 1573 h. But when I run the following I get about 1931 h. I may be messing something up here, but I want to be sure we get the yield calculations right.

```
import json
import PySAM.Pvwattsv7 as pvwatts
import PySAM.ResourceTools as tools
import PySAM.PySSC as pssc

# creating PySAM model with default info from json file (that doesn't have location info)
json_file = open("pvwatts_inputs.json")
pvwatts_dict = json.load(json_file)

# using the pvwatts model
pv_dat = pssc.dict_to_ssc_table(pvwatts_dict, "pvwattsv7")
json_file.close()
json_model = pvwatts.wrap(pv_dat)

solar_resource_file = tools.SAM_CSV_to_solar_data('weather_files/NV_36.083_-115.15_tmy.csv')
json_model.SolarResource.solar_resource_data = solar_resource_file

json_model.SystemDesign.gcr = 0.44
json_model.SystemDesign.array_type = 0
json_model.SystemDesign.tilt = 33
json_model.SystemDesign.dc_ac_ratio = 1.1

json_model.execute(0)

print(json_model.Outputs.kwh_per_kw)
```

sophiejandrews commented:
re the energy yield mismatch: for whatever reason, the longitude values on the csv I have to map coordinates to location names was mixed up for NV, NC and NY. but that’s fixed now and I built a tree (haven’t pushed to repo yet) that has values that match a single pysam run

sophiejandrews committed:
code review changes incorporated

sophiejandrews committed:
updated preset tree

sophiejandrews committed:
new preset tree based on new location mapping

mdeceglie commented on build-presets/MakePresetTree.py:
```diff
@@ -103,12 +102,11 @@
+                for f in weather_files:
+                    json_model.SolarResource.solar_resource_data = tools.SAM_CSV_to_solar_data(f)
+                    f = f.replace(weather_folder + '/', '') # remove folder name from file name
+                    id = f.split('_')[0] # get the id from the beginning of the string
```
`id` is another python built in we should avoid reassigning, maybe `location_id` or something?

mdeceglie commented on build-presets/get_weather_files.py:
```diff
@@ -14,19 +14,22 @@
 	lat = lat_list[i]
 	lon = lon_list[i]
 	state = states[i]
+	id = ids[i]
```
Another spot to avoid reassigning `id`

mdeceglie reviewed (commented):
One more minor suggestion, then I think we are good to go

sophiejandrews committed:
ILR js readability

sophiejandrews committed:
additional js comments

sophiejandrews committed:
updated comments

sophiejandrews committed:
license on BOS tree

sophiejandrews committed:
preset tree id renaming

sophiejandrews committed:
fixed typo
